src/Sarif/Visitors/ReformattingVisitor.cs

```diff
@@ -43,23 +45,129 @@ public override Run VisitRun(Run node)
+            {
+                goto Exit;
+            }
 
-            if (!Uri.TryCreate(key, UriKind.RelativeOrAbsolute, out uri))
+            // If the user hasn't specified code snippet insertion, there is no work to do
+            bool workToDo = false;
+            workToDo |= node.Region != null && _dataToInsert.Includes(OptionallyEmittedData.CodeSnippets);
+
+            if (!workToDo)
+            {
+                goto Exit;
+            }
+
+            // If we have an absolute URL specified as a file that 
+            // exist on disk, there is work to do
+            if (node.FileLocation.Uri.IsAbsoluteUri &&
+                node.FileLocation.Uri.LocalPath != null &&
+                File.Exists(node.FileLocation.Uri.LocalPath))
+            {
+                Region region = node.Region;
+                string fileText = GetFileText(node.FileLocation.Uri);
+                GetCompleteFirstLineAssociatedWithRegion(region, fileText);
+            }
+
+            Exit:
+            return base.VisitPhysicalLocation(node);
+        }
+
+        internal static void GetCompleteFirstLineAssociatedWithRegion(Region region, string fileText)
+        {
+            if (region.IsBinaryRegion)
+            {
+                return;
+            }
+
+            var lineIndex = new NewLineIndex(fileText);
+
+            region.Snippet = new FileContent();
+
+            int startLine;
+            LineInfo lineInfo;
+
+            if (region.StartLine == 0)
+            {
+                lineInfo = lineIndex.GetLineInfoForOffset(region.CharOffset);
+                startLine = lineInfo.LineNumber;
+            }
+            else
+            {
+                lineInfo = lineIndex.GetLineInfoForLine(region.StartLine);
+                startLine = region.StartLine;
+            }
+
+            int lineOffset = lineInfo.StartOffset;
+            int lineLength;
+
+            if (startLine == lineIndex.MaximumLineNumber)
+            {
+                // End of file. Compute length using file length
+                lineLength = fileText.Length - lineOffset;
+            }
+            else
+            {
+                // Not at end of file. Get the offset from the
+                // line after this one, then compute length from there.
+                lineInfo = lineIndex.GetLineInfoForLine(startLine + 1);
+                lineLength = lineInfo.StartOffset - lineOffset;
+            }
+
+            // Grab the line text, except for any trailing line break characters
+            region.Snippet.Text = fileText.Substring(lineOffset, lineLength).Trim();
+        }
+        
+
+        private string GetFileText(Uri uri)
+        {
+            _uriToFileTextMap = _uriToFileTextMap ?? new Dictionary<Uri, string>();
+            
+            if (!_uriToFileTextMap.TryGetValue(uri, out string fileText))
+            {
+                fileText = _uriToFileTextMap[uri] = File.ReadAllText(uri.LocalPath);
```

Comment:
I'm worried about this. You could end up sucking most of your code base into memory; if your code base is Windows, you'll regret it.

Reply:
Again, you can use `IFileSystem` here.

Reply:
Performance (speed) is already surfacing as an issue in the multitool. This could certainly represent a working set issue. It all depends on the scope of files referenced. We certainly don't want to reload files over and over again. Anyway, yes, we need some additional tuning, certainly before this code gets integrated into an actual service/pipeline.

---
In reply to: [202103169](https://github.com/Microsoft/sarif-sdk/pull/942#discussion_r202103169) [](ancestors = 202103169)